kustomization/kustomization_generator.go

```diff
@@ -0,0 +1,472 @@
+			extension := filepath.Ext(path)
+			if extension != ".yaml" && extension != ".yml" {
+				return nil
+			}
+
+			fContents, err := fs.ReadFile(path)
+			if err != nil {
+				return err
+			}
+
+			if _, err := rf.SliceFromBytes(fContents); err != nil {
+				return fmt.Errorf("failed to decode Kubernetes YAML from %s: %w", path, err)
+			}
+			paths = append(paths, path)
+			return nil
+		})
+		return paths, err
+	}
+
+	abs, err := filepath.Abs(dirPath)
+	if err != nil {
+		return unchangedAction, err
+	}
+
+	files, err := scan(abs)
+	if err != nil {
+		return unchangedAction, err
+	}
+
+	kfile := filepath.Join(dirPath, konfig.DefaultKustomizationFileName())
+	f, err := fs.Create(kfile)
+	if err != nil {
+		return unchangedAction, err
+	}
+	f.Close()
+
+	kus := kustypes.Kustomization{
+		TypeMeta: kustypes.TypeMeta{
+			APIVersion: kustypes.KustomizationVersion,
+			Kind:       kustypes.KustomizationKind,
+		},
+	}
+
+	var resources []string
+	for _, file := range files {
+		resources = append(resources, strings.Replace(file, abs, ".", 1))
+	}
+
+	kus.Resources = resources
+	kd, err := yaml.Marshal(kus)
+	if err != nil {
+		// delete the kustomization file
+		errf := CleanDirectory(dirPath, nil, createdAction)
+		return unchangedAction, fmt.Errorf("%v %v", err, errf)
+	}
+
+	return createdAction, os.WriteFile(kfile, kd, os.ModePerm)
+}
+
+func adaptSelector(selector *kustomize.Selector) (output *kustypes.Selector) {
+	if selector != nil {
+		output = &kustypes.Selector{}
+		output.Gvk.Group = selector.Group
+		output.Gvk.Kind = selector.Kind
+		output.Gvk.Version = selector.Version
+		output.Name = selector.Name
+		output.Namespace = selector.Namespace
+		output.LabelSelector = selector.LabelSelector
+		output.AnnotationSelector = selector.AnnotationSelector
+	}
+	return
+}
+
+// TODO: remove mutex when kustomize fixes the concurrent map read/write panic
+var kustomizeBuildMutex sync.Mutex
+
+// BuildKustomization wraps krusty.MakeKustomizer with the following settings:
+// - load files from outside the kustomization.yaml root
+// - disable plugins except for the builtin ones
+func BuildKustomization(fs filesys.FileSystem, dirPath string) (resmap.ResMap, error) {
+	// temporary workaround for concurrent map read and map write bug
+	// https://github.com/kubernetes-sigs/kustomize/issues/3659
+	kustomizeBuildMutex.Lock()
+	defer kustomizeBuildMutex.Unlock()
+
+	buildOptions := &krusty.Options{
+		LoadRestrictions: kustypes.LoadRestrictionsNone,
+		PluginConfig:     kustypes.DisabledPluginConfig(),
+	}
+
+	k := krusty.MakeKustomizer(buildOptions)
+	return k.Run(fs, dirPath)
+}
+
+// CleanDirectory removes the kustomization.yaml file from the given directory.
+func CleanDirectory(dirPath string, content []byte, action action) error {
+	kfile := filepath.Join(dirPath, konfig.DefaultKustomizationFileName())
+
+	if action == createdAction {
+		return os.Remove(kfile)
+	}
+
+	if action == unchangedAction {
+		err := os.WriteFile(kfile, content, 0644)
```

Comment:
I think it would be better to copy the origin file to `kustomization.yaml.original` before we inject the overrides, then this function would just restore the original.